test/Microsoft.IdentityModel.Tokens.Tests/CryptoProviderFactoryTests.cs

```diff
@@ -817,6 +829,144 @@ public static TheoryData<CryptoProviderFactoryTheoryData> ReleaseRsaKeyWrapProvi
+            Assert.Equal(1, signing._ReferenceCount);
+
+            var signing2 = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+            Assert.Equal(2, signing._ReferenceCount);
+            Assert.Equal(2, signing2._ReferenceCount);
+
+            cryptoProviderFactory.ReleaseSignatureProvider(signing2);
+
+            Assert.Equal(1, signing._ReferenceCount);
+
+            cryptoProviderFactory.ReleaseSignatureProvider(signing);
+
+            Assert.Equal(0, signing._ReferenceCount);
+        }
+
+        [Fact]
+        public void ReferenceCountingTest_NoCaching()
+        {
+            var context = new CompareContext($"{this}.ReferenceCountingTest_NoCaching");
+            var cryptoProviderFactory = new CryptoProviderFactory(CryptoProviderCacheTests.CreateCacheForTesting());
+            cryptoProviderFactory.CacheSignatureProviders = false;
+
+            var signing = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+            Assert.Equal(0, signing._ReferenceCount);
+
+            var signing2 = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+            Assert.Equal(0, signing._ReferenceCount);
+            Assert.Equal(0, signing2._ReferenceCount);
+
+            cryptoProviderFactory.ReleaseSignatureProvider(signing2);
+
+            Assert.Equal(0, signing._ReferenceCount);
+
+            cryptoProviderFactory.ReleaseSignatureProvider(signing);
+
+            Assert.Equal(0, signing._ReferenceCount);
+        }
+
+        [Fact]
+        public void ReferenceCountingTest_Caching()
+        {
+            var context = new CompareContext($"{this}.ReferenceCountingTest");
+            var cryptoProviderFactory = new CryptoProviderFactory(CryptoProviderCacheTests.CreateCacheForTesting());
+
+            var signing = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+            var signing2 = cryptoProviderFactory.CreateForSigning(Default.AsymmetricSigningKey, Default.AsymmetricSigningAlgorithm);
+
+            cryptoProviderFactory.CryptoProviderCache.TryRemove(signing2);
+
+            Assert.Null(signing.CryptoProviderCache);
+            Assert.Null(signing2.CryptoProviderCache);
+
+            cryptoProviderFactory.ReleaseSignatureProvider(signing2);
+            var disposeCalled = (bool)signing2.GetType().GetField("_disposed", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(signing2);
+            Assert.False(disposeCalled);
+            cryptoProviderFactory.ReleaseSignatureProvider(signing);
+            disposeCalled = (bool)signing.GetType().GetField("_disposed", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(signing);
+            Assert.True(disposeCalled);
+        }
+
+        [Fact]
+        public void ReferenceCountingTest_MultiThreaded()
```

Comment:
>public void ReferenceCountingTest_MultiThreaded() [](start = 8, length = 49)

Do you have a timing for this test? Would we be able to prove the same with fewer tasks?

Reply:
locally about 1 second, on the build machine about 6-7 seconds. Could we use less? Sure.